In app search improvements feature branch

bor0 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR is based on a feature branch that covers #51266, #51297, #51309, #51313, #51342, #51434.

We needed this feature branch to address issue dependencies.

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

1. Open each issue and test it
2. Make sure the search works, do some testing around it, and see if you can spot weird behavior

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

bor0 committed:
Add search results count to the in-app marketplace (#51266)

* Add searchResults to context

* Use setSearchResults in Content

* Add ribbons to the tabs

* Changelog

* Use setState as the function name

* Only show ribbon counts when there's an active search

* Refactor how 'setSearchResultsCount' is used (h/t @mcliwanow)

* Don't populate initial search results

* Unify css styling

raicem committed:
Marketplace: bring back the loading state (#51342)

* Marketplace: bring back the loading state

* Add changefile(s) from automation for the following project(s): woocommerce

---------

Co-authored-by: github-actions <github-actions@github.com>

bor0 committed:
Remove in-app marketplace Search results tab and unify results into existing tabs (#51297)

* Remove search results component and any references to it

* Persist current tab for searching, or default to extensions if tab is not set

* Persist term when switching across tabs

* Lint

* When a search is initiated, fetch all categories to keep the tab counts up to date.

The necessary filtering to display data to the current screen will be performed on the frontend.

* Apply correct colors to the tabs, as per design

* Beyond query.term, also rely on isLoading so that search result counts don't jump

* Address an issue when the user searches for something that returns no results in the business services tab

* Changelog

* Addressed :)

* Change key to category

* Fix category filter being broken

Whenever a category is requested, we need to do an additional request with the category param being the current category (overriding extensions/theme/business services).

Ideally the backend API would make a distinction between type (theme/extension/business service) and category, but this hack should do for now.

* Lint

* Remove unused variables h/t @KokkieH

* Lint

* Revert "Lint"

This reverts commit 0b2d2dca6de5087b3b0ae5599f8bb0114cebc4ab.

* Actually fix lint without introducing infinite loop

Reproducible at http://localhost:8080/wp-admin/admin.php?page=wc-admin&term=payments&tab=extensions&path=%2Fextensions&category=customer-service

* Show category selector even if there are no results displayed

* Update comment to be less misleading

* Query isn't used here

bor0 committed:
Merge branch 'trunk' into feature/in-app-search

KokkieH committed:
Update Marketplace search component (#51313)

* Update Search placeholder text

* Replace search component with one from @wordpress/components

* Make mobile search field font consistent with desktop

* Add changefile(s) from automation for the following project(s): woocommerce

* Handle import errors for SearchControl component

---------

Co-authored-by: github-actions <github-actions@github.com>

raicem committed:
Marketplace: update category selector (#51309)

* Marketplace: update category selector

Remove the dropdown on the desktop view and show all items, even if
overflowing. Added helper buttons to scroll to the right to show more.

* Add changefile(s) from automation for the following project(s): woocommerce

* Marketplace: remove category sroll helpers from tabindex

GitHub:
https://github.com/woocommerce/woocommerce/pull/51309/files#r1758448638

* Marketplace: Remove selectedTab reference from product.tsx

This is probably included due to the merge conflict

* Marketplace: tweak category scroll button narrower

---------

Co-authored-by: github-actions <github-actions@github.com>

bor0 committed:
Lint

bor0 committed:
Fix 2 lint errors

bor0 committed:
Fix another lint error (useMemo) h/t @KokkieH

andfinally commented:
Just testing https://github.com/woocommerce/woocommerce/pull/51266, and I'm only getting the count bubbles when I do a search.

I wonder if we need to announce `0` results? What do we think @poligilad-auto? Do we even need a bubble when there are no results? (The `0` on My subscriptions is also weird, but that must be a separate issue.)

<p>
<img width="456" alt="image" src="https://github.com/user-attachments/assets/6beeab18-a7bb-47de-b733-04034de6352c">
</p>

andfinally commented:
* For https://github.com/woocommerce/woocommerce/pull/51297, I get a count of 60 for Extensions and 5 for Themes from a search for "Bookings". `http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&term=Bookings`. 
* The API returns a total products count of 63 for my country `GB`. All of the items returned have `"type": "extension",` so I'm assuming there are no themes in those results.

KokkieH committed:
Add load more button in-app (#51434)

* Add additional fields returned by search API to marketplace types

Ensure components have access to additional fields

* Add LoadMoreButton component

* Only render Load More button if there are additional pages of results

* Fetch and display next page of results in Load More button is clicked

* Simplify renderContent function to have less repetition

- Hide load more button while fetching results

* Improve loading of new products

- Ensure keyboard focus goes to first new product after Load More is clicked

* Add changefile(s) from automation for the following project(s): woocommerce

* Add blank line to separate sections

* Set category param based on current tab when loading more products

* Improve busy-state screen reader text

Co-authored-by: Boro Sitnikovski <buritomath@gmail.com>

* Add missing dependency

* Move getProductType() function to functions.tsx

- Do not show load more button if isLoading state is true

---------

Co-authored-by: github-actions <github-actions@github.com>
Co-authored-by: Boro Sitnikovski <buritomath@gmail.com>

KokkieH commented:
> I get a count of 60 for Extensions
> The [endpoint](https://woocommerce.com/wp-json/wccom-extensions/1.0/search?term=Bookings&country=gb) returns a total products count of 63 for my country GB

The count only looks at the number of products in the filtered product list from the allProducts value, so the count is based on the number of products currently displayed, not the number of products available on WCCOM. But as the API now provides that value we can perhaps update how this is calculated?

bor0 commented:
Yes, after https://github.com/Automattic/woocommerce.com/pull/21678/files it seems like a good idea to rewrite how `setSearchResultsCount` is used. (Another dependency-issue we didn't consider 🙂)

andfinally commented:
Ah right, I get it. Yes, I'd update that, but it could wait to next release. I just wanted to spell out my results in case there was an issue.

bor0 committed:
Rework the values used with `setSearchResultsCount`

After https://github.com/Automattic/woocommerce.com/pull/21678/files we
get a `totalProducts` so we can re-use that.

Also remove setting the counts when paginating since we set them to the
total.

andfinally commented:
https://github.com/woocommerce/woocommerce/pull/51309 is good.

While testing this I noticed the themes cards overflow on mobile view. Added an [issue](21725-gh-Automattic/woocommerce.com) for this we can look at another time.

KokkieH commented:
@andfinally the theme card issue was fixed in https://github.com/woocommerce/woocommerce/pull/51187, which is already in trunk. If you update this branch against trunk you shouldn't see the issue any more.

andfinally commented:
https://github.com/woocommerce/woocommerce/pull/51313 is good. Thanks for thinking about accessibility! I noticed some unrelated accessibility issues – I created an [issue](21727-gh-Automattic/woocommerce.com), we can add to it as we find them.

https://github.com/woocommerce/woocommerce/pull/51342 looks good. I don't get the header on the "no results" message on the video on that PR. I also get the category tabs. But I guess that is a decision made during development.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/54ca7e66-5fcc-4262-be2e-6790a5f6f781">

> @andfinally the theme card issue was fixed in https://github.com/woocommerce/woocommerce/pull/51187, which is already in trunk. If you update this branch against trunk you shouldn't see the issue any more.

Awesome, so it does. Closing my issue for this.

KokkieH committed:
Merge branch 'trunk' into feature/in-app-search

KokkieH reviewed (commented):
On the Discover tab, if I do a search it switches me to the Extensions tab. However, if I do a search from the Subscriptions tab, it doesn't switch me. Is this intentional?

When searching, there's no screen reader announcement that the page content has been updated. It would be good if we can improve this before merging. [Here's a patch](https://github.com/user-attachments/files/17039719/search-a11y.patch) with a possible fix for this--doesn't work for the category filters, but it does work for search both with and without category filters set. With a category filter set, it should only announce the count for the current tab, else it announces the total across all tabs.

Aside from that everything tests well.

raicem commented:
> Hmm, @raicem fixed that the other day. I wonder if a regression slipped in when the category selector PR was rebased on the feature branch.

Thank you for catching this @andfinally and @KokkieH. It's indeed deleted now from this branch. I'll push a commit to add it back now

poligilad-auto commented:
Just jumping in here to say, please just notice that the [bubbles with search results](https://www.figma.com/design/2sXEw33oLyTZ7Mu3808sed/In-App-Marketplace?node-id=3916-72596&t=qWwmSnGMPYPW7sdO-1) should have different colors.

- Selected tab is black.
- Other tabs is gray.

The designs have a "0" as well, that's not a must, but I think the tab should be clickable and show the "No results" page.

andfinally commented:
> On the Discover tab, if I do a search it switches me to the Extensions tab. However, if I do a search from the Subscriptions tab, it doesn't switch me. Is this intentional?

This is maybe a consequence of the fact that the search doesn't cover My Subscriptions. Now that you mention it, it's not consistent. We can address it in a later release, if we don't get a chance to do it now.

> The designs have a "0" as well, that's not a must

I'm getting a red 0 bubble on My Subscriptions as well. We should not show a bubble there if there's nothing to announce – the red colour makes it especially distracting. I'm wondering if the logic that shows the 0 bubbles on the search tabs is somehow leaking on to the My Subscriptions tab.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/67d04cee-fdbd-4711-862c-755bffa2d67d">

bor0 committed:
Merge branch 'trunk' into feature/in-app-search

bor0 committed:
Add search complete announcement h/t @KokkieH

bor0 committed:
Show update count only if greater than 0 h/t @andfinally

bor0 committed:
Switch to Extensions tab if on My subscriptions when searching

bor0 commented:
> On the Discover tab, if I do a search it switches me to the Extensions tab. However, if I do a search from the Subscriptions tab, it doesn't switch me

Not intentional, nice catch. Fixed in 2b32972ec952d9788591082ef92c6d8d044bc9b8

> When searching, there's no screen reader announcement that the page content has been updated

This is a nice touch. Applied your patch in 21e675e2b9f5a5e95c77b21168fe16c7a6489da5

> Personally I'd be in favour of hiding bubbles if we have no count to show

This seems like a low risk. Applied in a93b8d8bcd9ce7e8b8d09b17775b6399ca1f962b

bor0 committed:
yoda

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -52,7 +69,100 @@ export default function Content(): JSX.Element {
+					return combinedProducts;
+				} );
+
+				speak( __( 'More products loaded', 'woocommerce' ) );
```
Nice!

andfinally reviewed (approved):
Beautiful! I checked the code for the changes from the last couple of days and retested them, all looks good. Thanks for the last-minute tweaks, which make it much better. 🙌

jorgemd24 commented:
Hey @bor0! Thanks for your work on this. I’ve been testing this feature (see here for context pb22l9-3kq-p2 )  and I noticed that clearing the search isn't working as expected. When I click the "X," the text clears, but the filtered results remain, so it's unclear how to return to the full list. I also noticed that the search only works after hitting "Enter," but I’d expect the results to filter as soon as I finish typing. Feel free to let me know if you'd like me to create an issue for this, or if I'm missing something!


https://github.com/user-attachments/assets/505a4fa3-3a07-4467-a9f0-4c089881eca2

bor0 commented:
@jorgemd24 thanks for noticing this. The PR didn't touch that part specifically, so it is a [known "issue"](https://github.com/woocommerce/woocommerce/blob/ce66b55/plugins/woocommerce-admin/client/marketplace/components/search/search.tsx#L58-L64). While it may be tricky to account for that use case*, feel free to create an issue and we can investigate.

*: There is no event we can use when the input resets, we can only rely on `input` or `click`, but we can't simply submit the search when the text is empty because a user can start typing, then change their mind (clear text) then start typing again.

jorgemd24 commented:
Thanks @bor0 for your explanation, I opened this issue: https://github.com/woocommerce/woocommerce/issues/51687
